File: packages/shared/src/api-response.ts
import { z } from "zod"
import type { ErrorCodeValue } from "./error-code.js"
import { ErrorCode } from "./error-code.js"

export const apiResponseSchema = <T extends z.ZodType>(dataSchema: T) =>
  z.object({
    code: z.number(),
    message: z.string(),
    data: dataSchema,
  })

export type ApiResponse<T> = {
  code: number
  message: string
  data: T
}

export function okResponse<T>(data: T, message = "ok"): ApiResponse<T> {
  return {
    code: ErrorCode.ok,
    message,
    data,
  }
}

export function failResponse<T>(code: ErrorCodeValue, message: string, data: T): ApiResponse<T> {
  return {
    code,
    message,
    data,
  }
}

File: apps/backend-spring/src/main/java/com/fullstack/springapi/common/ApiResponse.java
package com.fullstack.springapi.common;

public record ApiResponse<T>(int code, String message, T data) {

    public static <T> ApiResponse<T> ok(T data) {
        return new ApiResponse<>(0, "ok", data);
    }

    public static <T> ApiResponse<T> error(ErrorCode errorCode, T data) {
        return new ApiResponse<>(errorCode.code(), errorCode.message(), data);
    }

    public static <T> ApiResponse<T> error(ErrorCode errorCode, String message, T data) {
        return new ApiResponse<>(errorCode.code(), message, data);
    }
}
